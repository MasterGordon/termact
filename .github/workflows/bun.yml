name: bun CI

on:
  push:
    branches: [ "main" ]

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v3
    - uses: oven-sh/setup-bun@v1
      with:
        bun-version: latest
    - name: Install dependencies
      run: bun install
    - name: Run tests
      run: bun test
    - name: Run typecheck
      run: bun typecheck
