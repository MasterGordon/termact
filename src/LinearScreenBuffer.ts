import { logToFile } from "./lib/logToFile";

export class LinearScreenBuffer {
  private buffer: string[];
  private width: number;
  private height: number;

  constructor(width: number, height: number) {
    this.buffer = Array(height * width).fill(" ");
    this.width = width;
    this.height = height;
  }

  public get(x: number, y: number) {
    return this.buffer[y * this.width + x];
  }

  public set(x: number, y: number, value: string) {
    this.buffer[y * this.width + x] = value;
  }

  public getWidth() {
    return this.width;
  }

  public getHeight() {
    return this.height;
  }

  public resize(width: number, height: number) {
    logToFile("resize");
    if (width * height < this.buffer.length) {
      this.width = width;
      this.height = height;
      return;
    }

    this.buffer = Array(height * width).fill(" ");
    this.width = width;
    this.height = height;
  }

  public clear() {
    this.buffer = this.buffer.fill(" ");
  }

  public toString() {
    let res = "";
    for (let y = 0; y < this.height; y++) {
      res += this.buffer.slice(y * this.width, (y + 1) * this.width).join("");
    }
    return res;
  }
}
