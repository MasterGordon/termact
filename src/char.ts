const getChar = (c: string) => {
  return c.charCodeAt(0);
};

export const char = {
  ESC: "\u001b",
};
