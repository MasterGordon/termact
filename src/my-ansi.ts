export const csi = "\x1b[";
export const regexEscapedCsi = "\x1b\\[";

export const buttonsMask = {
  wheelUp: 64,
  wheelDown: 65,
  rightMb: 2,
  middleMb: 1,
  leftMb: 0,
} as const;

export const modifiersMask = {
  shift: 4,
  meta: 8,
  ctrl: 16,
};

const mouseEventRegex = new RegExp(
  `${regexEscapedCsi}<(\\d+);(\\d+);(\\d+)(M|m)$`
);

export const destructMouseEvent = (data: string) => {
  const match = mouseEventRegex.exec(data);
  if (match) {
    return {
      button: parseInt(match[1]),
      x: parseInt(match[2]),
      y: parseInt(match[3]),
      kind: match[4] as "M" | "m",
    };
  } else {
    return undefined;
  }
};

export const decodeButton = (code: number) => {
  const codeWithoutModifiers =
    code & ~(modifiersMask.shift | modifiersMask.meta | modifiersMask.ctrl);
  const button =
    Object.keys(buttonsMask).find(
      (k) => buttonsMask[k as keyof typeof buttonsMask] == codeWithoutModifiers
    ) ||
    ((codeWithoutModifiers & 32) == 32 // move
      ? "move"
      : undefined);

  if (button) {
    return {
      button: button as keyof typeof buttonsMask | "move",
      shift: !!(code & modifiersMask.shift),
      meta: !!(code & modifiersMask.meta),
      ctrl: !!(code & modifiersMask.ctrl),
    };
  } else {
    return undefined;
  }
};

export const isCtrlInput = (code: number) => code < 32;

export const decodeCtrlInput = (code: number) => {
  if (code == 0) return "@";
  else if (code == 27) return "[";
  else if (code == 28) return "\\";
  else if (code == 29) return "]";
  else if (code == 30) return "^";
  else if (code == 31) return "_";
  return String.fromCharCode(code + 64).toLowerCase();
};

const myAnsi = {
  // 1000 VT200 normal mouse tracking
  // 1006 SGR Extended coordinates
  // 1003 Any-event tracking
  enableMouseTracking: csi + "?1000;1006;1003h",
  disableMouseTracking: csi + "?1000;1006;1003l",
  enableAlternateBuffer: csi + "?1049h",
  disableAlternateBuffer: csi + "?1049l",
  hideCursor: csi + "?25l",
  showCursor: csi + "?25h",
  eraseInDisplay: csi + "2J",
} as const;

export default myAnsi;
