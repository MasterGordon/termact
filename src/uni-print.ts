function fixedHex(number: number, length: number) {
  var str = number.toString(16).toUpperCase();
  while (str.length < length) str = "0" + str;
  return str;
}

/* Creates a unicode literal based on the string */
function uniPrint(str: string): string {
  var i;
  var result = "";
  for (i = 0; i < str.length; ++i) {
    /* You should probably replace this by an isASCII test */
    if (str.charCodeAt(i) > 126 || str.charCodeAt(i) < 32)
      result += "\\u" + fixedHex(str.charCodeAt(i), 4);
    else result += str[i];
  }

  return result;
}

export default uniPrint;
