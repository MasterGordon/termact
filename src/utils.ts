import type { Frame } from "./types/Frame";

export const hasColorSupport = () => {
  return (
    process.env.COLORTERM === "truecolor" || process.env.COLORTERM === "24bit"
  );
};

export const hexToRgb = (hex: string) => {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  if (!result) {
    throw new Error(`Invalid hex color: ${hex}`);
  }
  return [
    parseInt(result[1], 16),
    parseInt(result[2], 16),
    parseInt(result[3], 16),
  ] as const;
};

export const rgbToHex = (r: number, g: number, b: number) => {
  return `#${r.toString(16).padStart(2, "0")}${g
    .toString(16)
    .padStart(2, "0")}${b.toString(16).padStart(2, "0")}`;
};

const color = (r: number, g: number, b: number, bg?: boolean) =>
  `\x1b[${bg ? "48" : "38"};2;${r};${g};${b}m`;

export const fg = (args: [number, number, number] | string) => {
  if (typeof args === "string") {
    return color(...hexToRgb(args));
  } else {
    const [r, g, b] = args;
    return color(r, g, b);
  }
};

export const bg = (args: [number, number, number] | string) => {
  if (typeof args === "string") {
    return color(...hexToRgb(args), true);
  } else {
    return color(...args, true);
  }
};

export const reset = () => "\x1b[0m";

export const bold = () => "\x1b[1m";

export const dim = () => "\x1b[2m";

export const italic = () => "\x1b[3m";

export const underline = () => "\x1b[4m";

export const inverse = () => "\x1b[7m";

export const hidden = () => "\x1b[8m";

export const cleanse = (text: string) => text.replace(/\u001b\[.+?m/g, "");

export const frame = (text: string, frame: Frame) => {
  const {
    vertical,
    horizontal,
    cornerTopLeft,
    cornerTopRight,
    cornerBottomLeft,
    cornerBottomRight,
  } = frame;

  const textWidth = Math.max(
    ...cleanse(text)
      .split("\n")
      .map((l) => l.length)
  );

  const modifiedText = text
    .split("\n")
    .map((l) => vertical + l.padEnd(textWidth) + vertical)
    .join("\n");
  return `${cornerTopLeft}${horizontal.repeat(textWidth)}${cornerTopRight}
${modifiedText}
${cornerBottomLeft}${horizontal.repeat(textWidth)}${cornerBottomRight}`;
};

export const center = (text: string) => {
  const textWidth = Math.max(...text.split("\n").map((l) => l.length));
  return text
    .split("\n")
    .map((l) => " ".repeat((textWidth - l.length) / 2) + l)
    .join("\n");
};

export const right = (text: string) => {
  const textWidth = Math.max(...text.split("\n").map((l) => l.length));
  return text
    .split("\n")
    .map((l) => l.padStart(textWidth))
    .join("\n");
};

export const moveCursor = (x: number, y: number) => {
  return `\x1b[${y};${x}H`;
};

// https://en.wikipedia.org/wiki/Grayscale#Luma_coding_in_video_systems
export const ntsc = (r: number, g: number, b: number) => {
  return r * 0.299 + g * 0.587 + b * 0.114;
};

// https://paulbourke.net/dataformats/asciiart/
const grayscaleCharacters =
  "$@B%8&WM#*oahkbdpqwmZO0QLCJUYXzcvunxrjft/\\|()1{}[]?-_+~<>i!lI;:,\"^`'. "
    .split("")
    .reverse();
export const getGrayscaleCharacter = (brightness: number) => {
  const index = Math.floor((brightness / 255) * grayscaleCharacters.length);
  return grayscaleCharacters[index] || " ";
};
