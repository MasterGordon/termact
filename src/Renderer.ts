import { Direction } from "yoga-layout";
import type Terminal from "./Terminal";
import type { Node } from "./objects/Box";
import { cleanse, style } from "./utils";

class Renderer {
  private terminal: Terminal;
  private terminalWidth: number;
  private terminalHeight: number;
  private root?: Node;

  constructor(terminal: Terminal) {
    this.terminal = terminal;
    this.registerEvents();
    [this.terminalWidth, this.terminalHeight] = terminal.getSize();
  }

  public setRoot(root: Node) {
    this.root = root;
    this.render();
  }

  public render() {
    this.terminal.clear();
    this.root!.node.calculateLayout(
      this.terminalWidth,
      undefined,
      Direction.LTR,
    );
    this.renderNode(this.root!);
  }

  private renderNode(node: Node) {
    const layout = node.node.getComputedLayout();

    const text = node.text || "";
    const lines = text.split("\n");

    const length = lines.length;
    for (let i = 0; i < layout.height - length; i++) lines.push("");

    lines.forEach((line, index) => {
      const cleansedLength = cleanse(line).length;
      line = style(node.style) + line;
      line += " ".repeat(layout.width - cleansedLength);
      this.terminal.printAt(layout.left, layout.top + index, line);
    });
    if (node.children.length > 0) {
      node.children.forEach((node) => this.renderNode(node));
    }
  }

  private registerEvents() {
    this.terminal.events.on("resize", (width, height) => {
      this.terminalWidth = width;
      this.terminalHeight = height;
      this.render();
    });
  }

  public destroy() {
    this.terminal.destroy();
  }
}

export default Renderer;
