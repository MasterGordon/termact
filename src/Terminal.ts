import { LinearScreenBuffer } from "./LinearScreenBuffer";
import EventEmitter from "./lib/EventEmitter";
import type { EventMap } from "./lib/EventEmitter";
import myAnsi, {
  csi,
  decodeButton,
  decodeCtrlInput,
  destructMouseEvent,
  isCtrlInput,
} from "./my-ansi";
import type { Implements } from "./types/Implements";
import uniPrint from "./uni-print";
import { moveCursor, splitCharacters } from "./utils";

interface Modifier {
  shift: boolean;
  meta: boolean;
  ctrl: boolean;
}

type Button = "leftMb" | "middleMb" | "rightMb" | "wheelUp" | "wheelDown";
type Kind = "down" | "up";

type TerminalEvents = Implements<
  {
    resize: [number, number];
    destroy: [];
    mouse: [number, number, Button, Kind, Modifier];
    mouseMove: [number, number];
    text: [string, Modifier];
  },
  EventMap
>;

interface TerminalOptions {
  stdin?: NodeJS.ReadableStream;
  stdout?: NodeJS.WritableStream;
  stderr?: NodeJS.WritableStream;
  alternateBuffer?: boolean;
  mouseTracking?: boolean;
  cursorHidden?: boolean;
  debugData?: boolean;
}

class Terminal {
  private stdin: NodeJS.ReadableStream;
  private stdout: NodeJS.WritableStream;
  private stderr: NodeJS.WritableStream;
  private options: TerminalOptions;
  private screenBuffer: LinearScreenBuffer;

  public events: EventEmitter<TerminalEvents>;

  constructor(options: TerminalOptions = {}) {
    this.stdin = options.stdin || process.stdin;
    this.stdout = options.stdout || process.stdout;
    this.stderr = options.stderr || process.stderr;
    this.options = options;
    const [terminalWidth, terminalHeight] = process.stdout.getWindowSize();
    this.screenBuffer = new LinearScreenBuffer(terminalWidth, terminalHeight);

    this.events = new EventEmitter<TerminalEvents>();

    this.init();
  }

  public write(text: string) {
    if (process.isBun) {
      Bun.write(Bun.stdout, text);
    } else {
      this.stdout.write(text);
    }
  }

  private registerEvents() {
    const dataListener = (data: string) => {
      if (data.length == 1) {
        if (isCtrlInput(data.charCodeAt(0))) {
          this.events.emit("text", decodeCtrlInput(data.charCodeAt(0)), {
            shift: false,
            meta: false,
            ctrl: true,
          });
        }
      }
      if (data.startsWith(csi)) {
        // Mouse tracking
        const destructured = destructMouseEvent(data);
        if (destructured) {
          const decoded = decodeButton(destructured.button);
          if (decoded) {
            if (decoded.button == "move") {
              this.events.emit("mouseMove", destructured.x, destructured.y);
            } else {
              const { button, ...modifier } = decoded;
              this.events.emit(
                "mouse",
                destructured.x,
                destructured.y,
                button,
                destructured.kind == "M" ? "down" : "up",
                modifier,
              );
            }
          }
        }
      }
      if (this.options.debugData) {
        this.stderr.write(uniPrint(data) + "\n");
      }
    };
    const processExitListener = () => {
      this.events.emit("destroy");
    };
    const resizeListener = () => {
      const [terminalWidth, terminalHeight] = process.stdout.getWindowSize();
      this.events.emit("resize", terminalWidth, terminalHeight);
      this.screenBuffer.resize(terminalWidth, terminalHeight);
    };

    process.on("exit", processExitListener);
    this.stdin.on("data", dataListener);
    this.stdout.on("resize", resizeListener);
    this.events.on("destroy", () => {
      this.stdin.off("data", dataListener);
      this.stdin.off("resize", resizeListener);
    });
    this.events.on("text", (text, modifier) => {
      if (modifier.ctrl && text == "c") {
        process.exit();
      }
    });
  }

  public printAt(x: number, y: number, text: string) {
    this.write(`${moveCursor(x + 1, y + 1) + text}\n`);
  }

  private init() {
    process.stdin.setRawMode(true);
    process.stdin.resume();
    process.stdin.setEncoding("utf8");

    if (this.options.alternateBuffer) {
      this.write(myAnsi.enableAlternateBuffer);
      this.events.on("destroy", () => {
        this.write(myAnsi.disableAlternateBuffer);
      });
    }

    if (this.options.mouseTracking) {
      this.write(myAnsi.enableMouseTracking);
      this.events.on("destroy", () => {
        this.write(myAnsi.disableMouseTracking);
      });
    }

    if (this.options.cursorHidden) {
      this.write(myAnsi.hideCursor);
      this.events.on("destroy", () => {
        this.write(myAnsi.showCursor);
      });
    }

    this.registerEvents();

    this.events.on("destroy", () => {
      process.stdin.setRawMode(false);
      process.stdin.pause();
      process.stdin.removeAllListeners();
    });
  }

  public destroy() {
    this.events.emit("destroy");
  }

  public clear() {
    this.write(myAnsi.eraseInDisplay);
  }

  public getSize() {
    return process.stdout.getWindowSize();
  }

  public printAtBuffer(x: number, y: number, text: string) {
    const lines = text.split("\n");
    for (let i = 0; i < lines.length; i++) {
      const chars = splitCharacters(lines[i]);
      for (let j = 0; j < chars.length; j++) {
        this.screenBuffer.set(x + j, y + i, chars[j]);
      }
    }
  }

  public flush() {
    this.printAt(0, 0, this.screenBuffer.toString());
  }
}

export default Terminal;
