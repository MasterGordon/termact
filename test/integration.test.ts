import { Terminal } from "@xterm/headless";
import { test, expect } from "bun:test";
import { frame, reset } from "../src/utils";
import { writeAsync } from "./xterm-utils";

test("integration", async () => {
  const terminal = new Terminal({
    allowProposedApi: true,
    convertEol: true,
  });

  terminal.resize(80, 25);
  await writeAsync(
    terminal,
    frame("My Test\nText\n..............", {
      vertical: "│",
      horizontal: "─",
      cornerTopLeft: "┌",
      cornerTopRight: "┐",
      cornerBottomLeft: "└",
      cornerBottomRight: "┘",
    })
  );

  const activeBuffer = terminal.buffer.active;

  let bufferString = [];
  for (let i = 0; i < activeBuffer.length; i++) {
    bufferString.push(activeBuffer.getLine(i)?.translateToString());
  }
  const screen = bufferString.map((s) => s?.trimEnd());

  expect(screen).toMatchSnapshot();
});
