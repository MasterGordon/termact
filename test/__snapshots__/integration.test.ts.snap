// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`integration 1`] = `
[
  "┌──────────────┐",
  "│My Test       │",
  "│Text          │",
  "│..............│",
  "└──────────────┘",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
  "",
]
`;
